### PR TITLE
Caching

### PR DESCRIPTION
**Pull request by 0xAaCE**
## Summary

Caching responses

## Trello Card

https://trello.com/c/lA0QKULz/16-cachear-las-respuestas

### CODE REVIEW

File: app/services/album.js [hunk 1]
```diff
@@ -1,27 +1,78 @@
...
-    .then(response => response.data)
-    .catch(e => {
-      throw errors.conectionError(e.message);
-    });
-};
+const albumLoader = new RedisDataLoader(
+  'album',
+  new DataLoader(
+    arrayId =>
+      Promise.resolve(
```
Comment (gonziesc) [inline app/services/album.js]: this should be promise.all, right?

File: app/services/album.js [hunk 2]
```diff
@@ -1,27 +1,78 @@
...
-    .get(endpoint)
-    .then(response => response.data)
-    .catch(e => {
-      throw errors.conectionError(e.message);
-    });
-};
+const allAlbumsLoader = new RedisDataLoader(
+  'allAlbums',
+  new DataLoader(arrayId =>
+    Promise.resolve(
```
Comment (gonziesc) [inline app/services/album.js]: same comment

File: app/services/album.js [hunk 3]
```diff
@@ -1,27 +1,78 @@
...
+        const endpoint = `${url}albums`;
+        return axios.get(endpoint);
+      })
+    )
+  ),
+  {
+    // caching here is a local in memory cache. Caching is always done
+    // to redis.
+    cache: true,
+    serialize: info => JSON.stringify(info.data)
```
Comment (gonziesc) [inline app/services/album.js]: why do you need stringify?

File: app/services/album.js [hunk 4]
```diff
@@ -1,27 +1,78 @@
...
+  }
+);
+
+const photosLoader = new RedisDataLoader(
+  'photos',
+  new DataLoader(arrayAlbumId =>
+    Promise.resolve(
```
Comment (gonziesc) [inline app/services/album.js]: same comment

File: app/services/album.js [hunk 5]
```diff
@@ -1,27 +1,78 @@
 const axios = require('axios');
 const { url } = require('../../config').common.api;
 const errors = require('../errors');
+const { user: User } = require('../models');
```
Comment (gonziesc) [inline app/services/album.js]: this file has too much logic now, how about if we split it?

**Review by gonziesc — CHANGES_REQUESTED** [review]

File: app/services/album.js [hunk 6]
```diff
@@ -37,10 +31,9 @@ const sortFunction = (array, orderBy) =>
     return 0;
   });
 
-exports.getAllAlbums = (offset, limit, filter, orderBy) => {
-  const endpoint = `${url}albums`;
-  return get(endpoint)
-    .then(response => response.data)
+exports.getAllAlbums = (offset, limit, filter, orderBy) =>
+  allAlbumsLoader()
+    .load('NaN')
```
Comment (gonziesc) [inline app/services/album.js:36]: why nan?

File: app/services/album.js (hunk 3 above)
Comment (0xAaCE) [reply app/services/album.js]: It is necessary because I just need stringify info.data

File: test/albums/cache.spec.js [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+      const fakeAlbumsProm = albumFactory.attributesMany(3).then(fakeAlbumsToMock => {
+        axios.setMockAlbums(fakeAlbumsToMock);
+        return fakeAlbumsToMock;
+      });
+      const fakePhotosProm = photoFactory.attributesMany(5).then(fakePhotosToMock => {
+        axios.setMockPhotos(fakePhotosToMock);
+        return fakePhotosToMock;
+      });
+      return Promise.all([fakeAlbumsProm, fakePhotosProm]).then(([fakeAlbums]) => {
+        let hrstart = moment();
```
Comment (gonziesc) [inline test/albums/cache.spec.js]: dont use let! make our vars immutable. use const and create another one

File: test/albums/cache.spec.js [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+        return query(getAlbum(fakeAlbums[0].id)).then(() => {
+          const hrendNoCached = moment().diff(hrstart);
+          hrstart = moment();
+          return query(getAlbum(fakeAlbums[0].id)).then(() => {
+            const hrendCached = moment().diff(hrstart);
+            return expect(hrendNoCached).toBeGreaterThan(hrendCached);
```
Comment (gonziesc) [inline test/albums/cache.spec.js]: I'd like to check the difference to be higger than some miliseconds (like 100). Just to check that there's no race conditions here

**Review by gonziesc — CHANGES_REQUESTED** [review]

**Review by gonziesc — APPROVED** [review]

**Pull request merged**